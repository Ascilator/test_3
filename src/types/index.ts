export type Country = {
  name: string;
  population: number;
  currencies: {
    code: string;
    name: string;
    symbol: string;
  }[];
};

export type ListType = {
  title: string;
  list: string[];
};

export type CountryReduced = ListType & {
  population: number;
};

export enum ListNames {
  countriesList = "countriesList",
  currencyList = "currencyList",
}

export enum ButtonStatus {
  NotActive,
  Active,
  Loading,
}
