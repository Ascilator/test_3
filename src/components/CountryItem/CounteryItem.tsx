"use client";

import { memo } from "react";
import cx from "classnames";

import { ButtonStatus, ListNames } from "@/types";
import { useListItem } from "@/hooks";

import { Button } from "../Button";

type CountryItemProps = {
  title: string;
  list: string[];
  listName: ListNames;
};

export const CountryItem: React.FC<CountryItemProps> = ({
  title,
  list,
  listName,
}) => {
  const { status, handleClick } = useListItem({ title, listName });

  let text = "loading";
  if (status === ButtonStatus.Active) {
    text = "on";
  }
  if (status === ButtonStatus.NotActive) {
    text = "off";
  }

  return (
    <div className="flex justify-between items-start py-3">
      <Button
        className={cx({
          "bg-lime-600 border-lime-600": status === ButtonStatus.Active,
          "bg-gray-600 border-gray-600": status === ButtonStatus.Loading,
        })}
        text={text}
        onClick={handleClick}
      />
      <div className="ml-5 mr-auto mt-3">{title}</div>
      <ul className="flex flex-col gap-3 text-right mt-3">
        {list.map((item) => (
          <div key={item}>{item}</div>
        ))}
      </ul>
    </div>
  );
};

export const MemoCountryItem = memo(CountryItem, (prevProps, nextProps) => {
  if (prevProps.title !== nextProps.title) {
    return false;
  }

  if (prevProps.list.length !== nextProps.list.length) {
    return false;
  }

  if (prevProps.listName !== nextProps.listName) {
    return false;
  }

  return true;
});
