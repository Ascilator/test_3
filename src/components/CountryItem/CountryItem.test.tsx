import React from "react";
import { ListNames } from "@/types";
import { CountryItem } from "./CounteryItem";
import { render } from "@testing-library/react";

describe("CountryItem test", () => {
  const componentProps = {
    title: "USD",
    list: ["USA", "Canada"],
    listName: ListNames.currencyList,
  };

  test("CountryItem is rendering", () => {
    const { container } = render(<CountryItem {...componentProps} />);
    expect(container).toMatchSnapshot();
  });

  test("CountryItem contains correct information", () => {
    const { getByText } = render(<CountryItem {...componentProps} />);

    expect(getByText("USD")).toBeTruthy();
    expect(getByText("USA")).toBeTruthy();
    expect(getByText("Canada")).toBeTruthy();
    expect(getByText("off")).toBeTruthy();
  });
});
