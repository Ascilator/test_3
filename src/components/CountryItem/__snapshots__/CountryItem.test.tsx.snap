// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryItem test CountryItem is rendering 1`] = `
<div>
  <div
    class="flex justify-between items-start py-3"
  >
    <button
      class="px-5 py-3 border border-cyan-500 bg-cyan-500 rounded-xl text-white "
    >
      off
    </button>
    <div
      class="ml-5 mr-auto mt-3"
    >
      USD
    </div>
    <ul
      class="flex flex-col gap-3 text-right mt-3"
    >
      <div>
        USA
      </div>
      <div>
        Canada
      </div>
    </ul>
  </div>
</div>
`;
